# Problem Set 5
### Written by Adithya Solai

In [38]:
import numpy as np
import pandas as pd

In this problem set, you will be exclusively working with data on MLB (Major League Baseball) teams from 1990 to 2015. You should have been given a baseball.csv file with the data in it. Make sure that .csv file is in the same file directory as this .ipynb file. Run the code in the cell below to load the data from this csv into a pandas DataFrame.

In [39]:
baseball_df = pd.read_csv("./datasets/baseball.csv")
baseball_df.head()

teamID  yearID  total_payroll franchID   W    G
0    BAL    1990        9680084      BAL  76  161
1    BOS    1990       20558333      BOS  88  162
2    CAL    1990       21720000      ANA  80  162
3    CHA    1990        9491500      CHW  94  162
4    CLE    1990       14487000      CLE  77  162

As seen above, this dataset tracks the # of wins each team got in a season (`W` column), the total # of games in that season (`G` column), and the amount of money in $ that the team spent on their player salaries that year (`total_payroll` column). This is done for all years from 1990 to 2015 (`yearID` column)

## Complete all tasks below.

### Task 1
**Drop the `teamID` column. The teamID changes throughout the years, as teams re-brand themselves with different cities/mascots/acronyms. The `franchId` (franchise ID) stays consistent throughout the years, so that column will be most helpful to our analysis.**

In [40]:
baseball_df.drop(columns='teamID', inplace=True)
baseball_df.head()

yearID  total_payroll franchID   W    G
0    1990        9680084      BAL  76  161
1    1990       20558333      BOS  88  162
2    1990       21720000      ANA  80  162
3    1990        9491500      CHW  94  162
4    1990       14487000      CLE  77  162

### Task 2
**Find the minimum and maximum value in the `yearID` column to ensure that this dataset is actually from 1990 to 2014.**

In [41]:
# There are many ways to do this problem, so it is OK if your solution differs.
print("Minimum yearID:", baseball_df['yearID'].min())
print("Maximum yearID:", baseball_df['yearID'].max())

Minimum yearID: 1990
Maximum yearID: 2014


### Task 3
**Create a new column called `win_pct` that represents the win percentage of each team for a particular season. Use the `W` and `G` column to easily create the new column. The `win_pct` column should be some `float` datatype, and all values in this column should be between 0 and 1 since it is a percentage.**

For example, the `win_pct` column for the first row in the dataset (`franchID`=BAL, `yearID`=1990) would be 76 / 161, or 0.47205.

In [42]:
baseball_df['win_pct'] = baseball_df['W'] / baseball_df['G']
baseball_df.head()

yearID  total_payroll franchID   W    G   win_pct
0    1990        9680084      BAL  76  161  0.472050
1    1990       20558333      BOS  88  162  0.543210
2    1990       21720000      ANA  80  162  0.493827
3    1990        9491500      CHW  94  162  0.580247
4    1990       14487000      CLE  77  162  0.475309

### Task 4
**Print the mean payroll for the New York Yankees (`franchID`=NYY) in the years 2000-2014. Do the same for the Oakland Athletics (`franchID`=OAK).**

In [43]:
baseball_df_2000_to_2014 = baseball_df[np.logical_and(baseball_df['yearID'] >= 2000, baseball_df['yearID'] <= 2014)]

baseball_df_2000_to_2014_nyy = baseball_df_2000_to_2014[baseball_df_2000_to_2014['franchID'] == 'NYY']
print("Mean Payroll for NYY from 2000-2014:", baseball_df_2000_to_2014_nyy['total_payroll'].mean())

baseball_df_2000_to_2014_oak = baseball_df_2000_to_2014[baseball_df_2000_to_2014['franchID'] == 'OAK']
print("Mean Payroll for OAK from 2000-2014:", baseball_df_2000_to_2014_oak['total_payroll'].mean())

Mean Payroll for NYY from 2000-2014: 180248411.2
Mean Payroll for OAK from 2000-2014: 55472697.2


### Task 5
**Find the team with the lowest average payroll from 1990-2015 (the entire dataset). Find the team with the highest average payroll from 1990-2015 as well.**

In [44]:
def find_avg_payroll(franchID):
  return baseball_df[baseball_df['franchID'] == franchID]['total_payroll'].mean()

# Instead of using .apply(), you can also just iterate through baseball_df['franchID'].unique() and use the 
# find_avg_payroll() function in each iteration, and store the average payrolls in a np.array or pd.Series,
# and fetch min() and max() after the for-loop.
baseball_df['avg_payroll'] = baseball_df['franchID'].apply(find_avg_payroll)
print('Team with Lowest Average Payroll: ', baseball_df.iloc[baseball_df['avg_payroll'].argmin()]['franchID'])
print('Lowest Average Payroll: ', baseball_df['avg_payroll'].min())
print('Team with Highest Average Payroll: ', baseball_df.iloc[baseball_df['avg_payroll'].argmax()]['franchID'])
print('Highest Average Payroll: ', baseball_df['avg_payroll'].max())

Team with Lowest Average Payroll:  PIT
Lowest Average Payroll:  37923528.72
Team with Highest Average Payroll:  NYY
Highest Average Payroll:  127869265.44


### Task 6
**Create a new column called `payroll_efficiency`, which is basically the `total_payroll` column divided by the `W` column. This will give a rough estimate of how efficient each team is at using their payroll to get wins.**

A lower `payroll_efficiency` value means that the team is more efficient since a low value is the result of a low numerator (low payroll spending) and a high denominator (high # of wins).

In [45]:
baseball_df['payroll_efficiency'] = baseball_df['total_payroll'] / baseball_df['W']
baseball_df.head()

yearID  total_payroll franchID   W    G   win_pct  avg_payroll  \
0    1990        9680084      BAL  76  161  0.472050  62797222.12   
1    1990       20558333      BOS  88  162  0.543210  94635165.16   
2    1990       21720000      ANA  80  162  0.493827  74239948.68   
3    1990        9491500      CHW  94  162  0.580247  66212115.72   
4    1990       14487000      CLE  77  162  0.475309  54564408.48   

   payroll_efficiency  
0       127369.526316  
1       233617.420455  
2       271500.000000  
3       100973.404255  
4       188142.857143

### Task 7
**Find the team with the lowest average `payroll_efficiency` from 1990-2015 to get a rough estimate of the most efficient team. Find the team with the highest average `payroll_efficiency` to find the team that is least efficient.**

In [46]:
# Use .apply() approach from before to creeate a new column 'avg_payroll_efficiency'

def find_avg_payroll_efficiency(franchID):
  return baseball_df[baseball_df['franchID'] == franchID]['payroll_efficiency'].mean()

# Instead of using .apply(), you can also just iterate through baseball_df['franchID'].unique() and use the 
# find_avg_payroll() function in each iteration, and store the average payrolls in a np.array or pd.Series,
# and fetch min() and max() after the for-loop.
baseball_df['avg_payroll_efficiency'] = baseball_df['franchID'].apply(find_avg_payroll_efficiency)

print('Team with Lowest Average Payroll Efficiency: ', baseball_df.iloc[baseball_df['avg_payroll_efficiency'].argmin()]['franchID'])
print('Lowest Average Payroll Efficiency: ', baseball_df['avg_payroll_efficiency'].min())
print('Team with Highest Average Payroll Efficiency: ', baseball_df.iloc[baseball_df['avg_payroll_efficiency'].argmax()]['franchID'])
print('Highest Average Payroll Efficiency: ', baseball_df['avg_payroll_efficiency'].max())

Team with Lowest Average Payroll Efficiency:  PIT
Lowest Average Payroll Efficiency:  521770.86931934603
Team with Highest Average Payroll Efficiency:  NYY
Highest Average Payroll Efficiency:  1374780.318820982
